### ISSUE TITLE
sort all sets before passing to tools

### ISSUE THREAD

**Issue by johnynek** [bazelbuild/rules_scala#531]
Some items in bazel are sets (srcs, deps, etc...) but end up in a file that is sent as an input somewhere. If the sets change order, we have to re-run the compiler.

This is a waste.

We should normalize all logical sets so it does not matter the order the user gives them. This will improve the cache hits.

cc @ianoc

**Comment by ittaiz** [comment]
By “if the sets change order” you mean due to Bazel changing implementation
or do you mean there’s no order guarantee?
Since I remember you saying depsets are ordered

**Comment by johnynek** [comment]
If users change the order of build files (i.e. they sort their build) it causes a rebuild.

**Comment by ittaiz** [comment]
Is this because depsets aren’t ordered?

**Comment by johnynek** [comment]
yes, it depsets were immutable sorted sets (e.g. an immutable red-black tree) everything would be fine.

But alas, no... We can't have nice things.

**Comment by ittaiz** [comment]
@laurentlb is this the desired state of things? Sounds like an easy caveat to stumble into when developing rules...

**Comment by ianoc** [comment]
@laurentlb the issue i think this refers to was that we template out the commands to feed into the compilers in their scripts. And these experimentally were not behaving stable (never built a repo of the problem alas, but adding a sort before templating removed the need to rebuild lots of targets after a clean + expunge and a rebuild of a repo with a remote cache). What order is preserved in the depset's? is it insertion order? or something else?

**Comment by laurentlb** [comment]
Depset ordering is documented here: https://docs.bazel.build/versions/master/skylark/depsets.html#order

**Comment by ianoc** [comment]
@laurentlb I've read that... specifically i guess what does `if this argument is omitted, the depset has the special default order, in which case there are no guarantees about the order of any of its elements.` imply? the order is maintained from construction but nothing else? or does that default order just mean no ordering?

**Comment by ianoc** [comment]
Also `but note that within each node direct elements have no order relative to children` too, i guess specifically for our case here that we to_list and want that to_list to be stable across different machines/rebuilds and so on as much as possible. I'm not sure from the docs it says that'll hold up so we need to sort after making the list?

**Comment by brandjon** [comment]
The `default` order does not specify any ordering guarantees aside from determinism, which is true of all things Skylark. (I'll clarify that sentence to mention determinism.)

In practice, however, it is implemented to be the same as `postorder`, except it won't complain when you try to combine it with depsets that use other orders (because it's not obligated to preserve another depset's order).

In all cases, the same inputs, specified orders, and sequences of union operations on depsets will produce the same output list. Changing machines shouldn't intrinsically change this directly (it's not a hash with a machine-dependent seed or anything). But if the other builds are running with some changed configuration then maybe that's affecting the contents.

**Comment by johnynek** [comment]
It's a shame that the depset is not really a "set" in the sense of equality: if you construct a list in different orders into a depset, they are not equivalent.

We only have reproducibility if the input order is exactly the same, which is not usually what people think of when they think of set equality.

Side note: depset as an immutable red-black tree seems like it would solve this (order is always sorted order). Maybe set could be reintroduced with this behavior.

**Comment by brandjon** [comment]
Not only are depsets unequal if they have the same contents in different orders, they are unequal if they have the same contents in the same order but are not an identical object in memory. Depsets follow reference equality semantics.

Depsets are explicitly not general-purpose sets. They are DAG-based collections with some order semantics, designed for efficient unioning. This makes them suitable for gathering dependencies across a build, but makes no attempt to provide the features of, say, a hash set.

An immutable red-black tree sounds interesting, but it's unlikely to be added anytime soon. I'm not very familiar with the data structure, but some of my concerns are the non-constant-time cost of unions, and reduced memory sharing of rebalanced parts of the tree.

It seems like anywhere that you're already reading the depset contents at analysis time, you can just add in a sort() to normalize the order. For places where the depset contents are read lazily at execution time (e.g. ctx.actions.args()), that's harder, and may warrant an API addition to support execution-time sorting.
